corehq/apps/accounting/invoicing.py

```diff
@@ -353,6 +326,163 @@ def create_wire_credits_invoice(self, items, amount):
         )
 
 
+class CustomerAccountInvoiceFactory(object):
+    """
+        This generates an invoice for a Customer Billing Account.
+    """
+    def __init__(self, date_start, date_end, account, recipients=None):
+        """
+        The Invoice generated will always be for the month preceding the
+        invoicing_date.
+        For example, if today is July 5, 2014 then the invoice will be from
+        June 1, 2014 to June 30, 2014.
+        """
+        self.date_start = date_start
+        self.date_end = date_end
+        self.account = account
+        self.recipients = recipients
+        self.customer_invoice = None
+
+    def create_invoice(self):
+        invoices = []
```

Comment:
It looks like the per-domain invoices are going to exist after the customer invoice is sent.  I think this is going to be a problem because the invoices will still be visible on pages like this one:

https://www.commcarehq.org/a/aangan/settings/project/billing/statements/

Ideally we wouldn't create any objects in postgres that aren't part of the customer invoice.

Reply:
Is it easier/preferable to not create all those invoices on the first go? Or just to delete the extra ones after the customer invoice has been sent?

Reply:
I'd recommend pulling out any shared functionality into functions and reusing the functions, instead of creating objects in the database.

I also think that calculating feature rate line items is not as simple as adding up line items from individual invoices.  For example, suppose 3 subscriptions share a customer software plan that includes 50 free users.  Project space A has 10 users, project space B has 55, and project space C has 60.  The total number of users is 10 + 55 + 60 = 125, so the number of users we'd charge for in the line item is 125 - 50 = 75.

If we calculated the subscription level invoices, the # users over the included amount would be 0, 5, 10, respectively, so we'd only end up charging for 0 + 5 + 10 = 15 users.